src/components/cc-product-list/products-controller.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * @typedef {import('./cc-product-list.types.js').ProductsCategory} ProductsByCategory
+ * @typedef {import('./cc-product-list').CcProductList} CcProductList
+ * @typedef {import('./cc-product-list.types.js').Product} Product
+ * @typedef {import('./cc-product-list.types.js').CategoryFilter} CategoryFilter
+ */
+
+export class ProductsController {
+
+  /**
+   * @param {CcProductList} host
+   */
+  constructor (host) {
+    /** @type {CcProductList} The host. */
+    this._host = host;
+
+    /** @type {ProductsByCategory[]} The initial products category list. */
+    this._productsByCategories = [];
+
+    /** @type {CategoryFilter[]} Categories and their current state.  */
+    this._categoriesFilters = [];
+
+    /** @type {string|null} Current category selected.  */
+    this._currentCategoryNameFilter = null;
+
+    /** @type {string|null} Current text filter.  */
+    this._currentTextFilter = '';
+  }
+
+  /**
+   * Returns a filtered list of the products with the filters applied.
+   *
+   * @returns {ProductsByCategory[]}
+   */
+  getFilteredProductsByCategories () {
+
+    const productsByCategoriesToggled = this._getProductsByCurrentCategory();
+    const productsByCategoriesTextFiltered = this._search(productsByCategoriesToggled, this._currentTextFilter);
+
+    // We either return:
+    // - the list of products filtered by the text and category
+    // - the list of products filtered only by the current category if the text filter is empty or null
+    return productsByCategoriesTextFiltered ?? productsByCategoriesToggled;
```

Comment:
nitpick:
- the `_search` method could have no parameters.
- the `_search` method should have all that logic.
- So that you could simplify that by `return this._search()`
- Having said that, do we really need the internal _search method? All the code in the _search method could be inlined here since the _search is used only here.

Reply:
:+1 Agreed with @pdesoyres-cc, let's remove `_search` and move the code here.